mcs/class/System/Mono.UnityTls/UnityTlsContext.cs

```diff
@@ -0,0 +1,370 @@
+				Unity.Debug.CheckAndThrow (errorState, "Failed to set list of supported ciphers", AlertDescription.HandshakeFailure);
+			}
+		}
+
+		public override bool ProcessHandshake ()
+		{
+			var errorState = UnityTls.NativeInterface.unitytls_errorstate_create ();
+			var result = UnityTls.NativeInterface.unitytls_tlsctx_process_handshake (m_TlsContext, &errorState);
+			if (errorState.code == UnityTls.unitytls_error_code.UNITYTLS_USER_WOULD_BLOCK)
+				return false;
+
+			// Not done is not an error if we are server and don't ask for ClientCertificate
+			if (result == UnityTls.unitytls_x509verify_result.UNITYTLS_X509VERIFY_NOT_DONE && IsServer && !AskForClientCertificate)
+				Unity.Debug.CheckAndThrow (errorState, "Handshake failed", AlertDescription.HandshakeFailure);
+			else
+				Unity.Debug.CheckAndThrow (errorState, result, "Handshake failed", AlertDescription.HandshakeFailure);
+
+			// .Net implementation gives the server a verification callback (with null cert) even if AskForClientCertificate is false.
+			// We stick to this behavior here.
+			if (IsServer && !AskForClientCertificate) {
+				if (!ValidateCertificate (null, null))
+					throw new TlsException (AlertDescription.HandshakeFailure, "Verification failure during handshake");
+			}
+
+			return true;
+		}
+
+		public override void FinishHandshake ()
+		{
+			// Query some data. Ignore errors on the way since failure is not crucial.
+			var errorState = UnityTls.NativeInterface.unitytls_errorstate_create ();
+			var cipherSuite = UnityTls.NativeInterface.unitytls_tlsctx_get_ciphersuite(m_TlsContext, &errorState);
+			var protocolVersion = UnityTls.NativeInterface.unitytls_tlsctx_get_protocol(m_TlsContext, &errorState);
+
+			m_Connectioninfo = new MonoTlsConnectionInfo () {
+				CipherSuiteCode = (CipherSuiteCode)cipherSuite,
+				ProtocolVersion = UnityTlsConversions.ConvertProtocolVersion(protocolVersion),
+				PeerDomainName = ServerName
+
+				// TODO:
+				// The following properties can be deducted from CipherSuiteCode.
+				// It looks though like as of writing no Mono implemention fills it out and there is also no mechanism that does that automatically
+				//
+				//CipherAlgorithmType
+				//HashAlgorithmType
+				//ExchangeAlgorithmType
+			};
+			m_IsAuthenticated = true;
+		}
+
+		[MonoPInvokeCallback (typeof (UnityTls.unitytls_tlsctx_write_callback))]
+		static private size_t WriteCallback (void* userData, byte* data, size_t bufferLen, UnityTls.unitytls_errorstate* errorState)
+		{
+			var handle = (GCHandle)(IntPtr)userData;
+			var context = (UnityTlsContext)handle.Target;
+			return context.WriteCallback (data, bufferLen, errorState);
+		}
+
+		private size_t WriteCallback (byte* data, size_t bufferLen, UnityTls.unitytls_errorstate* errorState)
+		{
+			try {
+				if (m_WriteBuffer == null || m_WriteBuffer.Length < bufferLen)
+					m_WriteBuffer = new byte[bufferLen];
+				Marshal.Copy ((IntPtr)data, m_WriteBuffer, 0, bufferLen);
```

Comment:
Does this need to be thread safe?

Reply:
There will be always only a single WriteCallback per context. So this should be safe :)